allow array-style simd in inline asm

Ezrashaw opened the pull request:
Required for [MCP#621](https://github.com/rust-lang/compiler-team/issues/621) to be implemented.

r? @workingjubilee

Ezrashaw committed:
allow array-style simd in inline asm

Ezrashaw commented on compiler/rustc_hir_analysis/src/check/intrinsicck.rs:
```diff
@@ -84,34 +106,16 @@ impl<'a, 'tcx> InlineAsmCtxt<'a, 'tcx> {
-                        })
+                    ty::Array(ty, len) => {
+                        let Some(len) = len.try_eval_target_usize(self.tcx, self.tcx.param_env(adt.did())) else {
+                            bug!("un-evaled const in SIMD type {ty} {len}");
```
Is this OK? I think so but I'm not 100% sure.

workingjubilee replied:
So, please use `span_bug!` at least.

workingjubilee replied:
Also, we shouldn't really be using even `span_bug!`. This will trigger an ICE, not a diagnostic. We should not be triggering an ICE just because we failed to get a value out of the to-be-evaluated constant parameter. We should trigger a diagnostic indicating the failure to evaluate and explain why.

Ezrashaw replied:
Oh wait, I'll just return `None` here. It matches what happens with the check for fields on `repr(simd)`.

compiler-errors replied:
I'm worried that this may be triggered with some creative usage of the `generic_const_exprs` feature. We should probably be just emitting a regular diagnostic here?

workingjubilee replied:
Basically, const parameters fail to evaluate for bullshit reasons that I don't personally understand all the time, but I do know they are going to happen on the regular and that even if this const eval path is hidden by some eager evaluation before this right _now_, the likelihood this code path will be exposed by a later code change is approximately 99.999% (something something "lazy normalization" probably), so we should just assume that it's possible for this to fail and default to assuming we should be giving a reasonable diagnostic.

workingjubilee replied:
race condition

Ezrashaw replied:
Yeah, I've fixed it.

scottmcm commented:
Good find!  And a great example of why there should be only one way to do things, so we don't have to think about this stuff :)

workingjubilee commented on compiler/rustc_hir_analysis/src/check/intrinsicck.rs:
```diff
@@ -84,34 +106,13 @@ impl<'a, 'tcx> InlineAsmCtxt<'a, 'tcx> {
                 let elem_ty = fields[FieldIdx::from_u32(0)].ty(self.tcx, substs);
+
                 match elem_ty.kind() {
-                    ty::Never | ty::Error(_) => return None,
```
This needs to remain on the same level, there are 40 LoC or so that this early return would skip.

Ezrashaw replied:
No? Just below the match is the following:
```rust
let Some(asm_ty) = asm_ty else {
    let msg = &format!("cannot use value of type `{ty}` for inline assembly");
    let mut err = self.tcx.sess.struct_span_err(expr.span, msg);
    err.note(
        "only integers, floats, SIMD vectors, pointers and function pointers \
         can be used as arguments for inline assembly",
    );
    err.emit();
    return None;
};
```

workingjubilee replied:
I consider the fact that the previous version of the code jumps out rather than even reaching that `else` and thus rather than emitting that error message to be significant.

workingjubilee replied:
...basically, `!` is *totally incoherent* for `repr(simd)` to begin with, so we should not be emitting the message regarding "but that's not a valid inline assembly input?" We should just be throwing up our hands and allowing the rest of rustc to emit a more general "what are you **doing**?"

Ezrashaw replied:
...and the whole point of `Ty::Error` is to stop more errors from being emitted. That was some great code from me lol.

workingjubilee commented on compiler/rustc_hir_analysis/src/check/intrinsicck.rs:
```diff
@@ -84,34 +106,13 @@ impl<'a, 'tcx> InlineAsmCtxt<'a, 'tcx> {
-                            64 => InlineAsmType::VecI64(fields.len() as u64),
-                            _ => unreachable!(),
-                        })
+                    ty::Array(ty, len)
```
Please recursively early return on the inner types of `ty::Array` that `elem_ty.kind()` also used to reject (shouldn't actually happen but I want to maintain consistency).

Ezrashaw replied:
The `simd_to_asm_type` already does this?

workingjubilee commented on tests/assembly/inline-asm-avx.rs:
```diff
@@ -0,0 +1,23 @@
+#[target_feature(enable = "avx")]
+pub unsafe fn convert(a: *const f32) -> Simd<f32, 8> {
+    let b: Simd<f32, 8>;
+    // CHECK: vbroadcastss
```
Please write the check so that it asserts that `ymm_reg` is in fact `ymm{something}`. I'm neutral on the other part of the instruction.

workingjubilee commented on compiler/rustc_hir_analysis/src/check/intrinsicck.rs:
```diff
@@ -44,6 +44,28 @@ impl<'a, 'tcx> InlineAsmCtxt<'a, 'tcx> {
 
+    fn simd_to_asm_type(&self, ty: Ty<'tcx>, size: u64) -> Option<InlineAsmType> {
+        match ty.kind() {
+            ty::Never | ty::Error(_) => return None,
```
Early returning at this level of control flow, being inside a different function, isn't the same as before. Just remove this and let the wildcard happen.

workingjubilee replied:
```suggestion
```

workingjubilee commented:
I am aware that some of my review, if followed-through on, may wind up with code that "ends up in the same place". My reasoning is that neither you nor me are the people doing the _most_ twiddling with our inline assembly implementation, so I think it is best if this code remains as similar to how it was as possible, even if I think the factored-out function is otherwise a good improvement and essential for this change.

Ezrashaw commented:
> I am aware that some of my review, if followed-through on, may wind up with code that "ends up in the same place".

Fair enough, I've updated the PR to align with that. Also I've moved the test to the `asm` subdirectory.

rust-log-analyzer commented:
The job **`mingw-check-tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/4771324069/jobs/8483187185) [(plain)](https://github.com/rust-lang/rust/commit/61ffb1d6dc16e2133c83fce6ea668af98e68bdf2/checks/12941727643/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
Prepare all required actions
Getting action download info
Download action repository 'actions/checkout@v3' (SHA:8e5e7e5ab8b370d6c329ec480221332ada57f0ab)
Download action repository 'rust-lang/simpleinfra@master' (SHA:4d706a8d448f000965ed9e883febfb36661202fe)
Complete job name: PR - mingw-check-tidy
git config --global core.autocrlf false
shell: /usr/bin/bash --noprofile --norc -e -o pipefail {0}
env:
  CI_JOB_NAME: mingw-check-tidy
---
Building wheels for collected packages: reuse
  Building wheel for reuse (pyproject.toml): started
  Building wheel for reuse (pyproject.toml): finished with status 'done'
  Created wheel for reuse: filename=reuse-1.1.0-cp310-cp310-manylinux_2_35_x86_64.whl size=180119 sha256=9fa76c45f3193f307e02ea67d1a48cfe7ef2b854a074b766938a88e046bc7887
  Stored in directory: /tmp/pip-ephem-wheel-cache-l4eysl7f/wheels/c2/3c/b9/1120c2ab4bd82694f7e6f0537dc5b9a085c13e2c69a8d0c76d
Installing collected packages: boolean-py, binaryornot, setuptools, reuse, python-debian, markupsafe, license-expression, jinja2, chardet
  Attempting uninstall: setuptools
    Found existing installation: setuptools 59.6.0
    Not uninstalling setuptools at /usr/lib/python3/dist-packages, outside environment /usr
---
Successfully built 3bb525c71e9c
Successfully tagged rust-ci:latest
Built container sha256:3bb525c71e9ca5ca71eb78fd577fe32d4deb5c14c20317378d71ad8b42007bc8
Uploading finished image to https://ci-caches.rust-lang.org/docker/eee721fc1ad06e35509f22d2087fb4acdb0a4d378bec3a191179769e2503f17c7f8f4292d7013976df3a24502e822c590708d80a62f327ca4849564995e4deb7
upload failed: - to s3://rust-lang-ci-sccache2/docker/eee721fc1ad06e35509f22d2087fb4acdb0a4d378bec3a191179769e2503f17c7f8f4292d7013976df3a24502e822c590708d80a62f327ca4849564995e4deb7 Unable to locate credentials
[CI_JOB_NAME=mingw-check-tidy]
[CI_JOB_NAME=mingw-check-tidy]
---
   Compiling cargo_metadata v0.15.3
   Compiling tidy v0.1.0 (/checkout/src/tools/tidy)
    Finished release [optimized] target(s) in 13.02s
fmt check
Diff in /checkout/compiler/rustc_hir_analysis/src/check/intrinsicck.rs at line 87:
                 let (size, ty) = match elem_ty.kind() {
                 let (size, ty) = match elem_ty.kind() {
                     ty::Array(ty, len) => {
-                        if let Some(len) = len.try_eval_target_usize(self.tcx, self.tcx.param_env(adt.did())) {
+                        if let Some(len) =
+                            len.try_eval_target_usize(self.tcx, self.tcx.param_env(adt.did()))
+                        {
                             (len, *ty)
                             return None;
                             return None;
Diff in /checkout/compiler/rustc_hir_analysis/src/check/intrinsicck.rs at line 99:
                 match ty.kind() {
                     ty::Never | ty::Error(_) => return None,
                     ty::Int(IntTy::I8) | ty::Uint(UintTy::U8) => Some(InlineAsmType::VecI8(size)),
-                    ty::Int(IntTy::I16) | ty::Uint(UintTy::U16) => Some(InlineAsmType::VecI16(size)),
-                    ty::Int(IntTy::I32) | ty::Uint(UintTy::U32) => Some(InlineAsmType::VecI32(size)),
-                    ty::Int(IntTy::I64) | ty::Uint(UintTy::U64) => Some(InlineAsmType::VecI64(size)),
-                    ty::Int(IntTy::I128) | ty::Uint(UintTy::U128) => Some(InlineAsmType::VecI128(size)),
+                    ty::Int(IntTy::I16) | ty::Uint(UintTy::U16) => {
+                        Some(InlineAsmType::VecI16(size))
+                    }
+                    ty::Int(IntTy::I32) | ty::Uint(UintTy::U32) => {
+                        Some(InlineAsmType::VecI32(size))
+                    }
+                    ty::Int(IntTy::I64) | ty::Uint(UintTy::U64) => {
+                        Some(InlineAsmType::VecI64(size))
+                    }
+                    ty::Int(IntTy::I128) | ty::Uint(UintTy::U128) => {
+                        Some(InlineAsmType::VecI128(size))
+                    }
                     ty::Int(IntTy::Isize) | ty::Uint(UintTy::Usize) => {
                         Some(match self.tcx.sess.target.pointer_width {
                             16 => InlineAsmType::VecI16(size),
Running `"/checkout/obj/build/x86_64-unknown-linux-gnu/rustfmt/bin/rustfmt" "--config-path" "/checkout" "--edition" "2021" "--unstable-features" "--skip-children" "--check" "/checkout/compiler/rustc_traits/src/dropck_outlives.rs" "/checkout/compiler/rustc_traits/src/implied_outlives_bounds.rs" "/checkout/compiler/rustc_traits/src/evaluate_obligation.rs" "/checkout/compiler/rustc_lexer/src/lib.rs" "/checkout/compiler/rustc_type_ir/src/structural_impls.rs" "/checkout/compiler/rustc_hir_analysis/src/check/intrinsicck.rs" "/checkout/compiler/rustc_hir_analysis/src/autoderef.rs" "/checkout/compiler/rustc_traits/src/normalize_projection_ty.rs"` failed.
If you're running `tidy`, try again with `--bless`. Or, if you just want to format code, run `./x.py fmt` instead.
```

</details>

workingjubilee commented on tests/assembly/asm/inline-asm-avx.rs:
```diff
@@ -0,0 +1,25 @@
+#[no_mangle]
+// CHECK-LABEL: convert:
+pub unsafe fn convert(a: *const f32) -> Simd<f32, 8> {
+    // CHECK: vbroadcastss (%rsi), {{%ymm[0-7]}}
```
I don't believe the specific scalar register is important.
```suggestion
    // CHECK: vbroadcastss (%{{[er][a-ds0-9][xpi0-9]?}}), {{%ymm[0-7]}}
```

workingjubilee commented:
This is a backwards-compatible change that purely adds new permitted behavior for unstable types and features. I have run the test suite and it Works On My Machine™.

@bors r+ rollup

bors commented:
:pushpin: Commit 85341836e0d4e44bc0b2772084cf5155a02d444a has been approved by `workingjubilee`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=workingjubilee 85341836e0d4e44bc0b2772084cf5155a02d444a -->
<!-- homu: {"type":"Approved","sha":"85341836e0d4e44bc0b2772084cf5155a02d444a","approver":"workingjubilee","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

matthiaskrgr commented:
@bors r- rollup=iffy
failed in a rollup https://github.com/rust-lang/rust/pull/110679#issuecomment-1518635640

Ezrashaw commented:
I've added the `force-host` UI test header; I believe this should fix the problem.

Nilstrieb commented:
I think we have shorter ways of making a test x86-only, don't we?

scottmcm commented on tests/assembly/asm/inline-asm-avx.rs:
```diff
@@ -0,0 +1,26 @@
+// assembly-output: emit-asm
+// compile-flags: --crate-type=lib
+// compile-flags: --target x86_64-unknown-linux-gnu
+// needs-llvm-components: x86
+// force-host
```
To @Nilstrieb's comment, I think you can do this as 
```suggestion
// only-x86_64
// ignore-sgx
```
(Why is `sgx` here?  I have no idea.  I just cargo-culted it from <https://github.com/rust-lang/rust/blob/master/tests/assembly/x86_64-floating-point-clamp.rs#L7>.)

Ezrashaw commented:
@scottmcm I've changed the UI test headers, CI passing. 

`compiletest` headers I will never understand lol.

workingjubilee commented:
I'd say "I should probably have caught that" except... yes, compiletest headers are many and confounding, sometimes.

workingjubilee commented:
@bors r+

bors commented:
:pushpin: Commit d31e8a499bdab8b70295580c288e5aa8d7f080dc has been approved by `workingjubilee`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=workingjubilee d31e8a499bdab8b70295580c288e5aa8d7f080dc -->
<!-- homu: {"type":"Approved","sha":"d31e8a499bdab8b70295580c288e5aa8d7f080dc","approver":"workingjubilee","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

workingjubilee commented:
Opened an issue for handling this better in the test suite or tidy checks: https://github.com/rust-lang/rust/issues/110751

bors commented:
:hourglass: Testing commit d31e8a499bdab8b70295580c288e5aa8d7f080dc with merge b72460fe46a873da0c40582636c26e7675e92288...
<!-- homu: {"type":"BuildStarted","head_sha":"d31e8a499bdab8b70295580c288e5aa8d7f080dc","merge_sha":"b72460fe46a873da0c40582636c26e7675e92288"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/4787125920/jobs/8515012515)
Approved by: workingjubilee
Pushing b72460fe46a873da0c40582636c26e7675e92288 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"workingjubilee","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/4787125920/jobs/8515012515"},"merge_sha":"b72460fe46a873da0c40582636c26e7675e92288"} -->

rust-timer commented:
Finished benchmarking commit (b72460fe46a873da0c40582636c26e7675e92288): [comparison URL](https://perf.rust-lang.org/compare.html?start=64bcb326516ef7490db46de88b87a4c0990097fe&end=b72460fe46a873da0c40582636c26e7675e92288&stat=instructions:u).

### Overall result: ❌ regressions - ACTION NEEDED


**Next Steps**: If you can justify the regressions found in this perf run, please indicate this with `@rustbot label: +perf-regression-triaged` along with sufficient written justification. If you cannot justify the regressions please open an issue or create a new PR that fixes the regressions, add a comment linking to the newly created issue or PR, and then add the `perf-regression-triaged` label to this PR.

@rustbot label: +perf-regression
cc @rust-lang/wg-compiler-performance



### [Instruction count](https://perf.rust-lang.org/compare.html?start=64bcb326516ef7490db46de88b87a4c0990097fe&end=b72460fe46a873da0c40582636c26e7675e92288&stat=instructions:u)
This is a highly reliable metric that was used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 2.1% | [0.9%, 2.9%] | 3     |
| Regressions ❌ <br /> (secondary)  | 6.4% | [5.2%, 8.1%] | 6     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | 2.1% | [0.9%, 2.9%] | 3     |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=64bcb326516ef7490db46de88b87a4c0990097fe&end=b72460fe46a873da0c40582636c26e7675e92288&stat=max-rss)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 3.4% | [3.4%, 3.4%] | 1     |
| Regressions ❌ <br /> (secondary)  | 2.6% | [2.6%, 2.6%] | 1     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | 3.4% | [3.4%, 3.4%] | 1     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=64bcb326516ef7490db46de88b87a4c0990097fe&end=b72460fe46a873da0c40582636c26e7675e92288&stat=cycles:u)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -    | -            | 0     |
| Regressions ❌ <br /> (secondary)  | 3.4% | [2.4%, 4.3%] | 5     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | -    | -            | 0     |
</details>

Mark-Simulacrum commented:
Cranelift-codegen and keccak are known to exhibit bimodality.
